src/spatial/flatten.jl

```diff
@@ -0,0 +1,90 @@
+using DiffEqJump, DiffEqBase, LightGraphs
+
+"""
+prob.u0 must be a Matrix with prob.u0[i,j] being the number of species i at site j
+"""
+function flatten(ma_jump, prob::DiscreteProblem, spatial_system, hopping_constants; kwargs...)
+    tspan = prob.tspan
+    u0 = prob.u0
+    if ma_jump === nothing
+        ma_jump = MassActionJump(Vector{typeof(tspan[1])}(),
+                             Vector{Vector{Pair{Int,eltype(u0)}}}(),
+                             Vector{Vector{Pair{Int,eltype(u0)}}}())
+    end
+    netstoch = ma_jump.net_stoch
+    reactstoch = ma_jump.reactant_stoch
+    rx_rates = ma_jump.scaled_rates
+    flatten(netstoch, reactstoch, rx_rates, spatial_system, u0, tspan, hopping_constants; scale_rates = false, kwargs...)
+end
+
+"""
+if hopping_constants is a matrix, assume hopping_constants[i,j] is species i, site j
+"""
+function flatten(netstoch::AbstractArray, reactstoch::AbstractArray, rx_rates::Vector, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Matrix{F}; kwargs...) where F <: Number
+    num_nodes = num_sites(spatial_system)
+    num_specs = size(u0, 1)
+    @assert size(hopping_constants) == size(u0) # hopping_constants[i,j] is species i, site j
+    hop_constants = Vector{Matrix{F}}(undef, size(hopping_constants, 2))
+    for site in 1:num_nodes
+        num_nbs = num_neighbors(spatial_system, site)
+        hop_constants[site] = reshape(repeat(hopping_constants[:,site], num_nbs), num_specs, num_nbs)
+    end
+    
+    flatten(netstoch, reactstoch, rx_rates, spatial_system, u0, tspan, hop_constants; kwargs...)
+end
+
+"""
+if reaction rates is a vector, assume reaction rates are equal across sites
+"""
+function flatten(netstoch::AbstractArray, reactstoch::AbstractArray, rx_rates::Vector, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; kwargs...) where F <: Number
+    num_nodes = num_sites(spatial_system)
+    rates = reshape(repeat(rx_rates, num_nodes), length(rx_rates), num_nodes)
+    flatten(netstoch, reactstoch, rates, spatial_system, u0, tspan, hopping_constants; kwargs...)
+end
+
+"""
+"flatten" the spatial jump problem. Return flattened DiscreteProblem and MassActionJump.
+"""
+function flatten(netstoch::Vector{R}, reactstoch::Vector{R}, rx_rates::Matrix{F}, spatial_system, u0::Matrix{Int}, tspan, hopping_constants::Vector{Matrix{F}}; scale_rates = true, kwargs...) where {R, F <: Number}
+    num_species = size(u0, 1)
+    num_nodes = num_sites(spatial_system)
+    num_rxs = length(reactstoch)
+    @assert length(hopping_constants) == size(u0, 2) == size(rx_rates, 2) == num_nodes
+    @assert length(netstoch) == length(reactstoch) == size(rx_rates, 1) == num_rxs
+    spec_CI = CartesianIndices((num_species, num_nodes))
+    spec_LI = LinearIndices((num_species, num_nodes))
+
+    sum_outdegrees = sum(num_neighbors(spatial_system, site) for site in 1:num_nodes)
+    num_jumps = num_species*(sum_outdegrees + num_rxs)
```

Comment:
Why is there a `num_species*num_rxs` term and not just `num_rxs`?

Reply:
Right. It should be `num_nodes * num_rxs` instead. Fixed.